### PR TITLE
Configure Content Security Policy script-src

### LINKED ISSUE #2086: Allow overriding Content security header
76.0.0 contains https://github.com/cloudfoundry/uaa/pull/1981/commits/72565fb56cd1f90af499119d32c891937f3c5a76 which disallows inline scripts.

On [GOV.UK PaaS ](https://login.cloud.service.gov.uk/login)we use javascript to improve some of the accessibility and on our customised passcode page to insert a copy code button that uses the clipboard api.

With the introduced change both of those things are no longer functional. 

Consider enabling a way to override the CSP settings.

### PR DESCRIPTION
**Pull request by jameshochadel**
https://github.com/cloudfoundry/uaa/commit/72565fb56cd1f90af499119d32c891937f3c5a76 disallowed inline scripts, but some users load scripts from other sources to add functionality to, for example, the login page. This change adds a new optional section to `uaa.yml` that allows custom script sources.

* The general approach is modeled on the `CorsFilter` and related `CorsConfiguration`.
* I wasn't sure where to put the new config class, so I co-located it with the `ContentSecurityPolicyFilter` itself.
* I couldn't find end-to-end tests for configuring filters from `uaa.yml`, so I just added a unit test for the new functionality. Let me know if integration tests for loading config exist and I'll add a test for the new configuration.

Addresses https://github.com/cloudfoundry/uaa/issues/2086.

### CODE REVIEW

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (strehle) [inline uaa/src/main/webapp/WEB-INF/spring-servlet.xml]: if you simply set self then CSP does not work. 'self' is wanted - I assume
-> ('''self''')

File: uaa/src/main/resources/uaa.yml
Comment (strehle) [inline uaa/src/main/resources/uaa.yml]: please remove it from here, examples should be placed in other places, e.g. cargo yml or uaa-release https://github.com/cloudfoundry/uaa-release/blob/develop/spec/compare/all-properties-set-uaa.yml

**Review by strehle — COMMENTED** [review]
do you consider a setting in uaa-release ?
https://github.com/cloudfoundry/uaa-release/pull/84

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (strehle) [inline uaa/src/main/webapp/WEB-INF/spring-servlet.xml]: I recommend to simply use csp and omit allowed e.g. uaa.yml from 

```
uaa.csp.script-src:
  - 'self'
  - 'unsafe-inline'
```

would place csp section in uaa-release
https://github.com/cloudfoundry/uaa-release/blob/develop/jobs/uaa/templates/config/uaa.yml.erb#L845

csp defines different section .e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy so currently we need only script-src , later we could add other in this section

**Comment by jameshochadel** [comment]
@strehle Thank you for the review! I've addressed the feedback. I matched the `csp` section to the `cors` section structure. And I will make a separate PR to `uaa-release` as you suggest.

**Comment by jameshochadel** [comment]
Sorry @strehle — I just re-requested your review, and it bumped all of the other reviewers off the PR for some reason. I don't have permission to add them back myself.

File: server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyConfiguration.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyConfiguration.java]: is this needed , I dont see a reason, we dont read json but this is in bean defined

File: server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyFilter.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyFilter.java]: you can also move the setting in the bean constructor, which is the best approach, so then the string can be final

File: server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyFilter.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyFilter.java:48]: can be called from construtor

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (strehle) [inline uaa/src/main/webapp/WEB-INF/spring-servlet.xml:251]: here you can define a constructor and set your property simply as input for constructor

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (strehle) [inline uaa/src/main/webapp/WEB-INF/spring-servlet.xml]: do you want to stay with scriptSrc ? I recommnd script-src to re-use same names from 
https://content-security-policy.com/script-src/

**Review by strehle — COMMENTED** [review]
for me change is OK , other maintainer should also check it

**Comment by strehle** [comment]
@jameshochadel did you see the comments, only wanted to know if you want this in next release. We currently have issues with the build pipeline but if these issues are fixed we create a new release

**Comment by jameshochadel** [comment]
@strehle Sorry for the delay; I saw your comments but was waiting for another maintainer review like you mentioned. I'll go through them and address them now!

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (jameshochadel) [reply uaa/src/main/webapp/WEB-INF/spring-servlet.xml]: I'm happy to do either. I was matching the CORS configuration casing, but it looks like other sections of `uaa.yml` use a variety of casing (camelCase in `ratelimit`, snake_case in `encryption`...). I'll match it to the hyphenated case as you suggest.

File: server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyConfiguration.java
Comment (jameshochadel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/security/web/ContentSecurityPolicyConfiguration.java]: I brought it over from the CORS configuration, so I'm not sure if it's needed. I'll remove it.

File: uaa/src/main/webapp/WEB-INF/spring-servlet.xml
Comment (strehle) [reply uaa/src/main/webapp/WEB-INF/spring-servlet.xml]: ok you can  stay with it, I know about these mixes... that if if you have open source, however I thought script-src might be easiert to understand - because of standard behind CSP - but I have no problem with the name.

**Comment by jameshochadel** [comment]
@strehle I believe I've addressed your feedback. Thank you again.

**Comment by strehle** [comment]
Ok, here I provided you a fix for the test and I think we should add a log , see my 2 commits from
https://github.com/cloudfoundry/uaa/commits/configure-csp-script-src

I will do log later, but please fix test in this PR and then we can merge it 

uaa-release is something you can do now

**Review by strehle — APPROVED** [review]
please fix test but LGTM

**Comment by jameshochadel** [comment]
@strehle The test has been updated. Sorry about that and thank you again.

**Comment by strehle** [comment]
sonar ok:
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2168

**Pull request merged**
